example/diamond.js

```diff
@@ -0,0 +1,178 @@
+			viewMatrixInv: { value: camera.matrixWorld },
+			resolution: { value: new THREE.Vector2( window.innerWidth, window.innerHeight ) }
+		},
+		vertexShader: /*glsl*/`
+            varying vec3 vWorldPosition;
+            varying vec3 vNormal;
+            uniform mat4 viewMatrixInv;
+            void main() {
+                vWorldPosition = (modelMatrix * vec4(position, 1.0)).xyz;
+                vNormal = (viewMatrixInv * vec4(normalMatrix * normal, 0.0)).xyz;
+                gl_Position = projectionMatrix * viewMatrix * modelMatrix * vec4(position, 1.0);
+            }
+    	`,
+		fragmentShader: /*glsl*/`
+            precision highp isampler2D;
+            precision highp usampler2D;
+
+            varying vec3 vWorldPosition;
+            varying vec3 vNormal;
+
+			uniform samplerCube envMap;
+            uniform float bounces;
+
+			${ shaderStructs }
+            ${ shaderIntersectFunction }
+
+			uniform BVH bvh;
+            uniform float ior;
+            uniform vec3 color;
+            uniform bool correctMips;
+            uniform mat4 projectionMatrixInv;
+            uniform mat4 viewMatrixInv;
+            uniform vec2 resolution;
+
+			void main() {
+                vec2 uv = gl_FragCoord.xy / resolution;
+                vec3 directionCamPerfect = (projectionMatrixInv * vec4(uv * 2.0 - 1.0, 0.0, 1.0)).xyz;
+
+			    directionCamPerfect = (viewMatrixInv * vec4(directionCamPerfect, 0.0)).xyz;
+                directionCamPerfect = normalize(directionCamPerfect);
+
+			    vec3 normal = vNormal;
+                vec3 rayOrigin = vec3(cameraPosition);
+                vec3 rayDirection = normalize(vWorldPosition - cameraPosition);
+                rayDirection = refract(rayDirection, normal, 1.0 / ior);
+                rayOrigin = vWorldPosition + rayDirection * 0.001;
+
+			    for(float i = 0.0; i < bounces; i++) {
+
+                    uvec4 faceIndices = uvec4( 0u );
+                    vec3 faceNormal = vec3( 0.0, 0.0, 1.0 );
+                    vec3 barycoord = vec3( 0.0 );
+                    float side = 1.0;
+                    float dist = 0.0;
+                    bvhIntersectFirstHit( bvh, rayOrigin, rayDirection, faceIndices, faceNormal, barycoord, side, dist );
+                    vec3 hitPos = rayOrigin + rayDirection * max(dist - 0.001, 0.0);
+                	// faceNormal *= side;
+                    vec3 tempDir = refract(rayDirection, faceNormal, ior);
+                    if (length(tempDir) != 0.0) {
+                        rayDirection = tempDir;
+                        break;
+                    }
+                    rayDirection = reflect(rayDirection, faceNormal);
+                    rayOrigin = hitPos + rayDirection * 0.01;
+
+                }
+
+                vec3 finalColor = textureGrad(envMap, rayDirection, dFdx(correctMips ? directionCamPerfect: rayDirection), dFdy(correctMips ? directionCamPerfect: rayDirection)).rgb * color;
```

Comment:
Can you explain what this snippet of code is doing? And what `textureGrad` and `correctMips` are used for?

Reply:
It's done to correct the mipmap level of the cubemap, to avoid ugly edges.